Heavy-Division/B78XH
# feat: electrical system

# Title
feat: electrical system






# Description 
This is a work in progress implementation of the electrical system in Rust. 

Problem: 
-Building aircraft systems can be an incredibly complex endeavor, and navigating the files in a .sln based repo will be difficult for open source based collaboration and integration. Based on the public repo itself, it's difficult to see the file organization and requires use of specialized IDEs llike Visual Studio to build and view the project structure. Heavy Division also has limited resources and developers as of writing this so writing an entirely custom aircraft system from scratch will be an extremely long endeavor. 

Solution: 
-The basic implementation of a commercial airliner system already exists, https://github.com/flybywiresim/a32nx/tree/master/src/systems At first glance, one might assume this is a broad copy-paste of the A320 Neo's electrical systems. However, Flybywire has done an excellent job of separating the basic electric principles and features found on all modern airliners from A320 Neo-specific behaviors. This allows anyone willing to open-source their projects to begin with this Rust template, and focus on writing the behaviors specific to their aircraft, as well as requiring less time to begin initial setup. Furthermore, Rust does not require IDE specific .sln behaviors, and thus viewing the systems structure from the repo or cloning to a wide range of IDEs will be much easier.

Some changes have been made to the template. Notably, any function or struct related to the bleed/pneumatic systems has been commented out, as the 787 is a no-bleed architecture. This leaves us with some basic tooling to work with in terms of engine and APU generators, Electrical Buses, and batteries as sources of potential to feed into various other systems. A separate directory labled 787_systems will need to be written to implement the aircraft specific behaviors, but much of the foundation such as simulation read/write, unit testing, and fundamental electric principles have been established. 



# Source material

B787 systems manual
![image](https://user-images.githubusercontent.com/52870481/197398110-b7441e35-7908-464d-a21b-575402c19118.png)



# Testing Instructions 
Build Test: Download the source code from this branch and extract the zip. and open the B78XH/src/systems/systems. Verify the systems build by running `cargo build` in the terminal. If you haven't downloaded Rust already, you can do so [here](https://www.rust-lang.org/tools/install)



# Test Results
![image](https://user-images.githubusercontent.com/52870481/197398315-decba60a-fbc7-4ae6-a274-89c41f3df393.png)




# Discord Username
HowNowBrownCrow#2591
